### PR TITLE
Add page maker app

### PR DESCRIPTION
**Pull request by baru64**

### CODE REVIEW

File: jolly-jellyfish/src/page_maker/fixtures/templates.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+- model: page_maker.template
+  pk: 1
+  fields:
+    name: hotdog
+    date_created: 2020-08-02 16:22:00+00:00
+    author: 1
+    style_sheet: styles/test.css
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/fixtures/templates.yaml:7]: Would it make more sense to have one of the included themes here?

File: jolly-jellyfish/src/page_maker/fixtures/templates.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+- model: page_maker.template
+  pk: 2
+  fields:
+    name: doghot
+    date_created: 2020-08-02 19:51:03.603472+00:00
+    author: 1
+    style_sheet: styles/test2.css
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/fixtures/templates.yaml:14]: As above.

File: jolly-jellyfish/src/page_maker/forms.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            'name', 'template_used', 'user_title', 'user_text_1', 'user_text_2', 'user_text_3',
+            'user_image_1', 'user_image_2', 'user_image_3', 'user_image_4'
+        ]
+
+
+class TemplateForm(forms.ModelForm):
+
+    class Meta:
+        model = Template
+        fields = ['name', 'style_sheet']
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/forms.py:25]: Would be cool if we could have a nicer widget for style_sheet, maybe with previews of the styles. Probably for another PR though.

File: jolly-jellyfish/src/page_maker/templates/page_maker/base.html [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      {% endif %}
+      <a href="{% url 'home' %}">home</a>
+    </header>
+
+    <main>
+      {% if error %}
+      <div>
+        <p>{{ error.text }}</p>
+      </div>
+      {% endif %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/base.html]: What sort of errors would these be? Could they be moved to after the content? Styling concerns...

File: jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html [hunk 5]
```diff
@@ -0,0 +1,13 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html]:
```suggestion
{% block content %}
<h2>Create a Template</h2>
<div>
```
Style finds first element and calls it a header.

File: jolly-jellyfish/src/page_maker/templates/page_maker/template_delete.html [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/template_delete.html]:
```suggestion
{% block content %}
<h2>Delete Template</h2>
<div>
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/page_maker/template_list.html [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+{% for template in templates %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/template_list.html]:
```suggestion
{% block content %}
<h2>Templates</h2>
{% for template in templates %}
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/page_maker/user_delete.html [hunk 8]
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/user_delete.html]:
```suggestion
{% block content %}
<h2>Delete Acount</h2>
<div>
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
+  username: {{ viewed_user.username }}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html]:
```suggestion
{% block content %}
username: {{ viewed_user.username }}
<div>
```
As above, styling concerns.

EDIT: Don't use this, see below.

File: jolly-jellyfish/src/page_maker/templates/page_maker/user_update.html [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/user_update.html]:
```suggestion
{% block content %}
<h2>Update Account</h2>
<div>
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_create.html [hunk 11]
```diff
@@ -0,0 +1,13 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_create.html]:
```suggestion
{% block content %}
<h2>Create Page</h2>
<div>
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html [hunk 12]
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html]:
```suggestion
{% block content %}
<h2>Delete Page</h2>
<div>
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html [hunk 13]
```diff
@@ -0,0 +1,16 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+{% for webpage in webpages %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html]:
```suggestion
{% block content %}
<h2>Top Pages</h2>
{% for webpage in webpages %}
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html [hunk 14]
```diff
@@ -0,0 +1,13 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html]:
```suggestion
{% block content %}
<h2>Update Page</h2>
<div>
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/registration/login.html [hunk 15]
```diff
@@ -0,0 +1,24 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+    {% if form.errors %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/registration/login.html]:
```suggestion
{% block content %}
    <h2>Login</h2>
    {% if form.errors %}
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/templates/registration/register.html [hunk 16]
```diff
@@ -0,0 +1,17 @@
+{% extends "page_maker/base.html" %}
+{% block content %}
+   <div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/registration/register.html]:
```suggestion
{% block content %}
   <h2>Sign up</h2>
   <div>
```
As above, styling concerns.

File: jolly-jellyfish/src/page_maker/tests.py [hunk 17]
```diff
@@ -1,3 +1,230 @@
...
+        self.assertEqual(user.email, 'foo@example.com')
+
+    def test_register_and_home(self):
+        # assumptions (aka what this tests):
+        #  - `/user/register` exists.
+        #  - it takes certain params.
+        #  - it also logs us in.
+        #  - `/` exists.
+        #  - it tells us our username.
+        #  - we use django's built-in auth system.
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/tests.py]: Might be better written as a docstring?

File: jolly-jellyfish/src/page_maker/tests.py [hunk 18]
```diff
@@ -1,3 +1,230 @@
...
+ADMIN_USERNAME = 'admin'
+ADMIN_PASSWORD = 'hard123passwd'
+
+
+class UserRegisterTest(TestCase):
+    def setUp(self):
+        # we create an account with the `USERNAME` and `PASSWORD` vars
+        params = {
+            'username': USERNAME,
+            'email': 'foo@example.com',
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/tests.py]: For consistency, email should probably be a constant also.

File: jolly-jellyfish/src/page_maker/tests.py [hunk 19]
```diff
@@ -1,3 +1,230 @@
...
+        }
+
+        self.client.post(reverse('register'), params)
+        self.client.login(username=USERNAME, password=PASSWORD)
+
+    def test_user_exists(self):
+        self.assertEqual(User.objects.count(), 1)
+        user = User.objects.get(pk=1)
+        self.assertEqual(user.username, USERNAME)
+        self.assertEqual(user.email, 'foo@example.com')
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/tests.py]: See above, email as a constant?

File: jolly-jellyfish/src/page_maker/tests.py [hunk 20]
```diff
@@ -1,3 +1,230 @@
...
+        self.client.login(username=USERNAME, password=PASSWORD)
+
+    def test_template_create(self):
+        # what this assumes:
+        # - when you create a page the
+        #   available templates are
+        #   passed in. and in a certain
+        #   way.
+        # - posting `/templates/new`
+        #   creates a specified template
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/tests.py]: Perhaps better written as a docstring?

File: jolly-jellyfish/src/page_maker/views.py [hunk 21]
```diff
@@ -1,3 +1,242 @@
...
+
+
+class UserDetailView(DetailView):
+    model = User
+    context_object_name = 'viewed_user'
+    template_name = 'page_maker/user_detail.html'
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        # user = get_object_or_404(User, pk=self.kwargs.get('pk'))
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/views.py]: Useless code?

File: jolly-jellyfish/src/page_maker/views.py [hunk 22]
```diff
@@ -1,3 +1,242 @@
...
+        user = self.get_object()
+        if self.request.user == user:
+            return True
+        return False
+
+
+class WebpageCreateView(LoginRequiredMixin, FormView):
+    template_name = 'page_maker/webpage_create.html'
+    form_class = WebpageForm
+    # success_url = '/pages'
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/views.py]: Useless code?

File: jolly-jellyfish/src/page_maker/views.py [hunk 23]
```diff
@@ -1,3 +1,242 @@
...
+
+    # TODO validate and sanitize text and images
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/views.py]: How much of this does Django do for us?

File: jolly-jellyfish/src/page_maker/views.py [hunk 24]
```diff
@@ -1,3 +1,242 @@
...
+        parent_page = get_object_or_404(Webpage, name=self.kwargs.get('pagename'))
+        form.instance.parent_page = parent_page
+        form.save()
+        return super().form_valid(form)
+
+    def get_success_url(self):
+        pagename = self.kwargs.get('pagename')
+        return reverse_lazy('webpage-detail', kwargs={'pagename': pagename})
+    
+    # redirect to detail view
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/views.py]: Better written inside the function/as a docstring?

**Review by Artemis21 — CHANGES_REQUESTED** [review]
Looks good. Several suggestions adding titles to pages, mainly for styling.

File: jolly-jellyfish/src/page_maker/fixtures/templates.yaml (hunk 1 above)
Comment (A5rocks) [reply jolly-jellyfish/src/page_maker/fixtures/templates.yaml:7]: Pretty sure these files don't exist to begin with:

https://github.com/Artemis21/summer-code-jam-2020/pull/25#discussion_r467231618

File: jolly-jellyfish/src/page_maker/fixtures/users.yaml [hunk 25]
```diff
@@ -0,0 +1,30 @@
+- model: auth.user
+  pk: 1
+  fields: # password: hard123passwd
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/fixtures/users.yaml]: Hm maybe we want to doublespace this comment?

File: jolly-jellyfish/src/page_maker/fixtures/users.yaml [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+    last_name: ''
+    email: admin@example.com
+    is_staff: true
+    is_active: true
+    date_joined: 2020-08-06 18:14:02+00:00
+    groups: []
+    user_permissions: []
+- model: auth.user
+  pk: 2
+  fields: # password: foobar < xyzzy
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/fixtures/users.yaml]: Same as above

File: jolly-jellyfish/src/page_maker/forms.py [hunk 27]
```diff
@@ -0,0 +1,36 @@
+from django import forms
+from django.contrib.auth.forms import UserCreationForm
+
+from .models import Comment, Webpage, Template, User
+
+
+class CommentForm(forms.ModelForm):
+
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/forms.py]: An unnecessary linebreak?

File: jolly-jellyfish/src/page_maker/forms.py [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+    class Meta:
+        model = Comment
+        fields = ['title', 'content']
+
+
+class WebpageForm(forms.ModelForm):
+
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/forms.py]: Same as above?

File: jolly-jellyfish/src/page_maker/forms.py [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+    class Meta:
+        model = Webpage
+        fields = [
+            'name', 'template_used', 'user_title', 'user_text_1', 'user_text_2', 'user_text_3',
+            'user_image_1', 'user_image_2', 'user_image_3', 'user_image_4'
+        ]
+
+
+class TemplateForm(forms.ModelForm):
+
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/forms.py]: Same as above...

File: jolly-jellyfish/src/page_maker/templates/page_maker/base.html [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+
+    {% if webpage %}
+      <title>{{ webpage.user_title }}</title>
+    {% else %}
+      <title>Early Internet Page Maker</title>
+    {% endif %}
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/page_maker/base.html]: Pretty sure we aren't indenting inside if blocks...

File: jolly-jellyfish/src/page_maker/templates/page_maker/base.html [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+  <body>
+    <header>
+      Hello!
+      {% if user.is_authenticated %}
+        <a href="{% url 'user-detail' pk=user.id %}"> {{ user.username }}</a>,
+        <a href="{% url 'logout' %}">logout</a>
+      {% else %}
+        <a href="{% url 'login' %}">Login</a>
+        <a href="{% url 'register' %}">Register</a>
+      {% endif %}
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/page_maker/base.html]: Same as above

File: jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html (hunk 9 above)
Comment (A5rocks) [reply jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html]: Shouldn't there be a header too?

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+  {% endfor %}
+</div>
+<div>
+  <form method="POST" enctype="multipart/form-data" action="{% url 'comment-create' pagename=webpage.name %}">
+    {% csrf_token %}
+      <div>
+        {{ comment_form.as_p }}
+        <button type="submit">Add comment</button>
+      </div>
+    </form>
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html]: Extra indent.

File: jolly-jellyfish/src/page_maker/templates/registration/login.html [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+         <tr>
+             <td>{{ form.password.label_tag }}</td>
+             <td>{{ form.password }}</td>
+         </tr>
+         </table>
+ 
+         <input type="submit" value="login" />
+         <input type="hidden" name="next" value="{{ next }}" />
+     </form>
+{% endblock %}
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/registration/login.html]: Weirdly indented? As in, everything inside the block should have 1 level less of indentation.

File: jolly-jellyfish/src/page_maker/templates/registration/login.html [hunk 34]
```diff
@@ -0,0 +1,24 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+    {% if form.errors %}
+        <p>Your username and password didn't match. Please try again.</p>
+    {% endif %}
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/registration/login.html]: This probably shouldn't be indented extra.

File: jolly-jellyfish/src/page_maker/templates/registration/login.html [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+         </tr>
+         <tr>
+             <td>{{ form.password.label_tag }}</td>
+             <td>{{ form.password }}</td>
+         </tr>
+         </table>
+ 
+         <input type="submit" value="login" />
+         <input type="hidden" name="next" value="{{ next }}" />
+     </form>
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/registration/login.html]: All these are indented with 1 space too much :O

... What am I doing with my life to be scanning the space count like that in a couple places :///

File: jolly-jellyfish/src/page_maker/templates/registration/register.html [hunk 36]
```diff
@@ -0,0 +1,17 @@
...
+                <button type="submit"> Sign Up </button>
+            </div>
+        </form>
+    <div>
+        <small>
+            Already Have An Account?  <a href="{% url 'login' %}"> Sign In </a>
+        </small>
+    </div>
+    </div>
+{% endblock content %}
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/registration/register.html]: Weird indentation? As in: there's a level of indentation that shouldn't be there in the content block.

File: jolly-jellyfish/src/page_maker/templates/registration/register.html [hunk 37]
```diff
@@ -0,0 +1,17 @@
+{% extends "page_maker/base.html" %}
+{% block content %}
+   <div>
+       <form method="POST">
+           {% csrf_token %}
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/registration/register.html]: In most places we have this line 1 indent lower (to be the same level as the form before)

**Review by A5rocks — CHANGES_REQUESTED** [review]
I very much skimmed through this one (as Artemis has already gone through and I've gone through the constituent PRs), just pointed out a couple things I found.

Two things:
  1) We'll want to format html at the end for consistency's sake.
  2) This will be a PAIN to resolve merge conflicts with tameTNT's PR... You and him took different directions a bunch.

File: jolly-jellyfish/src/page_maker/fixtures/templates.yaml (hunk 1 above)
Comment (Artemis21) [reply jolly-jellyfish/src/page_maker/fixtures/templates.yaml:7]: Oh, ok. @baru64 can you explain why?

**Comment by Artemis21** [comment]
As @A5rocks has pointed out, it would be good to lint/beautify the html, and perhaps also the yaml. Does anyone know of good linters for django-html or yaml?

File: jolly-jellyfish/src/page_maker/fixtures/templates.yaml (hunk 1 above)
Comment (baru64) [reply jolly-jellyfish/src/page_maker/fixtures/templates.yaml:7]: We only test the behavior of view, what objects does it return, what is the status code and when we check if there is an image or file in a page we only check if there is a link to it, we don't check if you can download the image.
To use actual files we can change the MEDIA_ROOT for tests to something like /test_media/ but there's not point in this unless we want to test downloading.

File: jolly-jellyfish/src/page_maker/templates/page_maker/base.html (hunk 4 above)
Comment (baru64) [reply jolly-jellyfish/src/page_maker/templates/page_maker/base.html]: I just dropped it in before creating views, but I will remove it since it's not used.

File: jolly-jellyfish/src/page_maker/views.py (hunk 23 above)
Comment (baru64) [reply jolly-jellyfish/src/page_maker/views.py]: It probably handles sanitizing both text and images, I removed the comment.

File: jolly-jellyfish/src/page_maker/templates/page_maker/base.html (hunk 30 above)
Comment (Artemis21) [reply jolly-jellyfish/src/page_maker/templates/page_maker/base.html]: Maybe we should though

File: jolly-jellyfish/src/page_maker/templates/registration/login.html [hunk 38]
```diff
@@ -0,0 +1,23 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+{% if form.errors %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/registration/login.html]:
```suggestion
{% block content %}
<h2>Log in</h2>
{% if form.errors %}
```
Still missing a header.

File: jolly-jellyfish/src/page_maker/templates/registration/register.html [hunk 39]
```diff
@@ -0,0 +1,17 @@
+{% extends "page_maker/base.html" %}
+{% block content %}
+<div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/registration/register.html]:
```suggestion
{% block content %}
<h2>Sign Up</h2>
<div>
```
As above.

**Review by Artemis21 — COMMENTED** [review]
Header still missing in a couple of files.

File: jolly-jellyfish/src/page_maker/fixtures/templates.yaml (hunk 1 above)
Comment (Artemis21) [reply jolly-jellyfish/src/page_maker/fixtures/templates.yaml:7]: Oh, of course, I thought this was what we discussed about including some articles default.

File: jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html (hunk 9 above)
Comment (Artemis21) [reply jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html]: Oh, you're right
```suggestion
{% block content %}
<h2>Username: {{ viewed_user.username }}</h2>
<div>
```

File: jolly-jellyfish/src/page_maker/templates/registration/login.html [hunk 40]
```diff
@@ -1,24 +1,24 @@
...
+{% csrf_token %}
+  <table>
+  <tr>
+    <td>{{ form.username.label_tag }}</td>
+    <td>{{ form.username }}</td>
+  </tr>
+  <tr>
+    <td>{{ form.password.label_tag }}</td>
+    <td>{{ form.password }}</td>
+  </tr>
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/templates/registration/login.html]:
```suggestion
    <tr>
      <td>{{ form.username.label_tag }}</td>
      <td>{{ form.username }}</td>
    </tr>
    <tr>
      <td>{{ form.password.label_tag }}</td>
      <td>{{ form.password }}</td>
    </tr>
```

very minor change though

File: jolly-jellyfish/src/page_maker/tests.py [hunk 41]
```diff
@@ -163,17 +167,20 @@ def setUp(self):
...
+           way.
+         - POST-ing `/templates/new`
+           creates a specified template
+         - The console theme exists.
+        """
 
         # example stylesheet
-        style = io.StringIO('body { display: hidden; }')
         template_name = 'blahblahblah'
+        style = str(Path(settings.BASE_DIR) / 'static' / 'themes' / 'console.css')
```
Comment (A5rocks) [inline jolly-jellyfish/src/page_maker/tests.py:183]: We might want to revert this change.

**Review by A5rocks — COMMENTED** [review]
LGTM, just a really minor nitpick and something that we should make a separate issue for

**Review by A5rocks — APPROVED** [review]

**Comment by A5rocks** [comment]
These are probably beyond the scope of the PR:
 - https://github.com/Artemis21/summer-code-jam-2020/pull/28#discussion_r467342421
 - https://github.com/Artemis21/summer-code-jam-2020/pull/28#discussion_r467483310

Let's put them into another PR so we can finally merge this one :^)

File: jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html [hunk 42]
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Create a template</h2>
+<div>
+  <form method="POST" enctype="multipart/form-data">
+  {% csrf_token %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html:6]: Should this line be further indented?

File: jolly-jellyfish/src/page_maker/templates/page_maker/template_delete.html [hunk 43]
```diff
@@ -0,0 +1,15 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Delete a template</h2>
+<div>
+  <form method="POST">
+  {% csrf_token %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/template_delete.html:6]: As above

File: jolly-jellyfish/src/page_maker/templates/page_maker/user_delete.html [hunk 44]
```diff
@@ -0,0 +1,15 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Delete an account</h2>
+<div>
+  <form method="POST">
+  {% csrf_token %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/user_delete.html:6]: As above

File: jolly-jellyfish/src/page_maker/templates/page_maker/user_update.html [hunk 45]
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Update an account</h2>
+<div>
+  <form method="POST">
+  {% csrf_token %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/user_update.html:6]: As above.

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_create.html [hunk 46]
```diff
@@ -0,0 +1,14 @@
...
+
+{% block content %}
+<h2>Create a webpage</h2>
+<div>
+  <form method="POST" enctype="multipart/form-data">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <button type="submit">Create page</button>
+    </div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_create.html]: What is the purpose of this div?

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html [hunk 47]
```diff
@@ -0,0 +1,15 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Delete a webpage</h2>
+<div>
+  <form method="POST">
+  {% csrf_token %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html:6]: Should this line be further indented?

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html [hunk 48]
```diff
@@ -0,0 +1,15 @@
...
+    <div>
+      {{ form.as_p }}
+      <p>Delete page <b>{{webpage.name}}?</b></p>
+      <button type="submit">Confirm</button>
+    </div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html]: What is the purpose of this div?

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html [hunk 49]
```diff
@@ -0,0 +1,30 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<div><h1>{{ webpage.user_title }}</h1></div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html]:
```suggestion
<h1><span><span>{{ webpage.user_title }}</span></span></h1>
```
This will have the custom theme, right?

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html [hunk 50]
```diff
@@ -0,0 +1,30 @@
...
+  </div>
+  {% endfor %}
+</div>
+<div>
+  <form method="POST" enctype="multipart/form-data" action="{% url 'comment-create' pagename=webpage.name %}">
+    {% csrf_token %}
+    <div>
+      {{ comment_form.as_p }}
+      <button type="submit">Add comment</button>
+    </div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html]: Unnecessary div?

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html [hunk 51]
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Update a webpage</h2>
+<div>
+  <form method="POST">
+  {% csrf_token %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html:6]: Indent?

File: jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html [hunk 52]
```diff
@@ -0,0 +1,14 @@
...
+    <div>
+      {{ form.as_p }}
+      <button type="submit">Update</button>
+    </div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html]: Unnecessary div?

File: jolly-jellyfish/src/page_maker/templates/registration/login.html [hunk 53]
```diff
@@ -0,0 +1,24 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Log in</h2>
+{% if form.errors %}
+<p>Your username and password didn't match. Please try again.</p>
+{% endif %}
+
+<form method="post" action="{% url 'login' %}">
+{% csrf_token %}
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/registration/login.html]: Indent?

File: jolly-jellyfish/src/page_maker/templates/registration/register.html [hunk 54]
```diff
@@ -0,0 +1,18 @@
+{% extends "page_maker/base.html" %}
+{% block content %}
+<h2>Sign Up</h2>
+<div>
+  <form method="POST">
+    {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <button type="submit"> Sign Up </button>
+    </div>
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/templates/registration/register.html]: Unnecessary div?

File: jolly-jellyfish/src/page_maker/tests.py [hunk 55]
```diff
@@ -1,3 +1,237 @@
...
+USERNAME = 'haha69420'
+PASSWORD = 'foobar < xyzzy'
+EMAIL = 'foo@example.com'
+
+ADMIN_USERNAME = 'admin'
+ADMIN_PASSWORD = 'hard123passwd'
+
+
+class UserRegisterTest(TestCase):
+    def setUp(self):
```
Comment (Artemis21) [inline jolly-jellyfish/src/page_maker/tests.py:21]: This name seems out of style, or is it required by something else?

File: jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html (hunk 42 above)
Comment (A5rocks) [reply jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html:6]: I guess? We'll have to edit tameTNT's PR to have this too though.

File: jolly-jellyfish/src/page_maker/tests.py (hunk 55 above)
Comment (A5rocks) [reply jolly-jellyfish/src/page_maker/tests.py:21]: It's required by `unittest`.

**Review by Artemis21 — APPROVED** [review]
